include/scip2/walltime.h

```diff
@@ -38,13 +40,34 @@ class Walltime
       initialized_ = true;
     }
 
-    if (time_device < (1 << DEVICE_TIMESTAMP_BITS) / 2 &&
-        (1 << DEVICE_TIMESTAMP_BITS) / 2 < time_device_prev_)
+    if (detectDeviceTimeJump(time_device))
+    {
+      logger::warn() << "Device time jumped." << std::endl;
+      if (walltime_device_base_ >= (1 << DEVICE_TIMESTAMP_BITS))
+      {
+        time_device_prev_ = (1 << DEVICE_TIMESTAMP_BITS) - time_device;
+        return walltime_device_base_ - time_device_prev_;
+      }
+    }
+
+    if (time_device < middle_bits_ &&
+        middle_bits_ < time_device_prev_)
+    {
       walltime_device_base_ += 1 << DEVICE_TIMESTAMP_BITS;
+    }
+
     time_device_prev_ = time_device;
 
     return walltime_device_base_ + time_device;
   }
+
+  bool detectDeviceTimeJump(const uint32_t& time_device)
```

Comment:
I think this is something like `detect...Underflow` rather than `detectDeviceTimeJump`.